server/api/batch_inventory.py

```diff
@@ -223,6 +241,151 @@ def parse_vote(vote: str):
+            entry_names = unzip_files(cvr_file, working_directory)
+            file_names = [
+                entry_name
+                for entry_name in entry_names
+                if entry_name.endswith(".csv")
+                and not entry_name.startswith(".")
+                and not entry_name.startswith("__")
+            ]
+            cvr_file.close()
+
+            cvr_and_ballots_files = separate_ess_cvr_and_ballots_files(
+                working_directory, file_names
+            )
+            primary_cvr_file, ballots_files = (
+                cvr_and_ballots_files["cvr_file"],
+                cvr_and_ballots_files["ballots_files"],
+            )
+
+            cvr_number_to_batch = {}
+            for ballots_file in ballots_files.values():
+                _, headers, rows = read_ess_ballots_file(ballots_file)
+                header_indices = get_header_indices(headers)
+                for row_index, row in enumerate(rows):
+                    cvr_number = column_value(
+                        row,
+                        "Cast Vote Record",
+                        row_index + 1,
+                        header_indices,
+                        required=True,
+                    )
+                    batch = column_value(
+                        row, "Batch", cvr_number, header_indices, required=True
+                    )
+                    cvr_number_to_batch[cvr_number] = batch
+
+            validate_comma_delimited(primary_cvr_file)
+            cvr_csv = csv.reader(primary_cvr_file, delimiter=",")
+            headers = next(cvr_csv)
+            header_indices = get_header_indices(headers)
+
+            contest_names = [contest.name for contest in contests]
+            missing_contest_names = set(contest_names) - set(headers)
+            if len(missing_contest_names) != 0:
+                raise UserError(
+                    f"CVR file is missing contest names: {', '.join(missing_contest_names)}"
+                )
+
+            for row_index, row in enumerate(cvr_csv):
+                for contest in contests:
+                    cvr_number = column_value(
+                        row,
+                        "Cast Vote Record",
+                        row_index + 1,
+                        header_indices,
+                        required=True,
+                    )
+                    choice_name = column_value(
+                        row,
+                        contest.name,
+                        row_index + 1,
+                        header_indices,
+                        required=False,
+                    )
+
+                    if cvr_number not in cvr_number_to_batch:
+                        raise UserError(
+                            f"Unable to find batch for CVR number {cvr_number} in ballots files"
+                        )
+                    batch = cvr_number_to_batch[cvr_number]
+                    batch_key: BatchKey = ("", batch)
```

Comment:
The `BatchKey` type allows for a `(tabulator_id, batch_id)` pair, but the ES&S CVR files that we've seen thus far don't include particularly useful tabulator info. The lack of a tabulator ID is handled just fine when generated ballot manifests and candidate-totals-by-batch files in [this previous commit](https://github.com/votingworks/arlo/pull/1923/commits/6d5f3f8fec5344ba82e9b01bd8c6eb4834f6a600)